pymemcache/codecs.py

```diff
@@ -0,0 +1,170 @@
+
+    def __init__(self, pickle_version=0):
+        """
+        Init
+
+        :param int pickle_version: Pickle version to use (from python).
+          Use `-1` to use the highest supported at runtime.
+          Deserialization is not affected by this parameter.
+
+          A forewarning with `0` (the default): If somewhere in your value lies
+          a slotted object, ie defines `__slots__`, even if you do not include
+          it in your pickleable state via `__getstate__`, python will raise:
+            ```
+            TypeError: a class that defines __slots__ without defining
+            __getstate__ cannot be pickled
+            ```
+        """
+        if pickle_version is not None:
+            self.pickle_version = pickle_version
+
+    def from_python(self, key, value):
+        """
+        Serialize a python object.
+
+        :param str|unicode key: Key
+        :param str|unicode value: Value
+        :return tuple[str, int]: tuple(value, flags)
+        """
+        flags = 0
+        value_type = type(value)
+
+        # Check against exact types so that subclasses of native types will be
+        # restored as their native type
+        if value_type is bytes:
+            pass
+
+        elif value_type is six.text_type:
+            flags |= self.FLAG_TEXT
+            value = value.encode('utf8')
+
+        elif value_type is int:
+            flags |= self.FLAG_INTEGER
+            value = "%d" % value
+
+        elif six.PY2 and value_type is long_type:
+            flags |= self.FLAG_LONG
+            value = "%d" % value
+
+        else:
+            flags |= self.FLAG_PICKLE
+
+            output = BytesIO()
+            pickler = pickle.Pickler(output, self.pickle_version)
+            pickler.dump(value)
+            value = output.getvalue()
+
+        return value, flags
+
+    def to_python(self, key, value, flags):
+        """
+        Deserialize a value into a python object.
+
+        :param str|unicode key: Key
+        :param str|unicode value: Value
+        :param int flags: Bitflag containing flags used to specify how to
+        deserialize this object.
+        :return object: Deserialized python object.
+        """
+        if flags == 0:
+            return value
+
+        elif flags & self.FLAG_TEXT:
+            return value.decode('utf8')
+
+        elif flags & self.FLAG_INTEGER:
+            return int(value)
+
+        elif flags & self.FLAG_LONG:
+            if six.PY3:
+                return int(value)
+            else:
+                return long_type(value)
+
+        elif flags & self.FLAG_PICKLE:
+            try:
+                buf = BytesIO(value)
+                unpickler = pickle.Unpickler(buf)
+                return unpickler.load()
+            except Exception as exc:
+                # This includes exc as a string for troubleshooting as well as
+                # providing a trace.
+                log.exception('Could not depickle value (len=%d): %s',
```

Comment:
Because from the *initial* line, you know what the exc was that caused it. This helps when an exception storm hits and you have N of them in your log interspersed.